datafusion/optimizer/src/optimizer.rs

```diff
@@ -288,11 +288,17 @@ impl Optimizer {
             log_plan(&format!("Optimizer input (pass {i})"), &new_plan);
 
             for rule in &self.rules {
-                let result = self.optimize_recursively(rule, &new_plan, config);
-
+                let result =
+                    self.optimize_recursively(rule, &new_plan, config)
+                        .and_then(|plan| {
+                            plan.map(|p| {
+                                assert_schema_is_the_same(rule.name(), &new_plan, &p)
+                                    .map(|_| Some(p))
+                            })
+                            .unwrap_or(Ok(None))
+                        });
```

Comment:
Here is another way of writing this logic which I think might be clearer:

```suggestion
                let result =
                    self.optimize_recursively(rule, &new_plan, config)
                    .and_then(|plan| {
                        if let Some(plan) = &plan {
                            assert_schema_is_the_same(rule.name(), &new_plan, &plan)?;
                        }
                        Ok(plan)
                    });
```

Reply:
Thanks @alamb , I've added your suggestion. Defy looks much better this way 🥇